File: library/src/main/java/ru/terrakok/cicerone/android/FragmentNavigator.java
/*
 * Created by Konstantin Tskhovrebov (aka @terrakok)
 */

package ru.terrakok.cicerone.android;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;

import java.util.ArrayDeque;
import java.util.Iterator;

import ru.terrakok.cicerone.Navigator;
import ru.terrakok.cicerone.commands.Back;
import ru.terrakok.cicerone.commands.BackTo;
import ru.terrakok.cicerone.commands.Command;
import ru.terrakok.cicerone.commands.Forward;
import ru.terrakok.cicerone.commands.Replace;

/**
 * {@link Navigator} implementation based on the fragments.
 * <p>
 * {@link BackTo} navigation command will return to the root if
 * needed screen isn't found in the screens chain.
 * To change this behavior override {@link #backToUnexisting(String)} method.
 * </p>
 * <p>
 * {@link Back} command will call {@link #exit()} method if current screen is the root.
 * </p>
 */
public abstract class FragmentNavigator implements Navigator {
    private FragmentManager fragmentManager;
    private int containerId;
    private ArrayDeque<String> localStackCopy;

    /**
     * Creates FragmentNavigator.
     *
     * @param fragmentManager fragment manager
     * @param containerId     id of the fragments container layout
     */
    public FragmentNavigator(FragmentManager fragmentManager, int containerId) {
        this.fragmentManager = fragmentManager;
        this.containerId = containerId;
    }

    /**
     * Override this method to setup fragment transaction {@link FragmentTransaction}.
     * For example: setCustomAnimations(...), addSharedElement(...) or setReorderingAllowed(...)
     *
     * @param command             current navigation command. Will be only {@link Forward} or {@link Replace}
     * @param currentFragment     current fragment in container
     *                            (for {@link Replace} command it will be screen previous in new chain, NOT replaced screen)
     * @param nextFragment        next screen fragment
     * @param fragmentTransaction fragment transaction
     */
    protected void setupFragmentTransaction(Command command,
                                                     Fragment currentFragment,
                                                     Fragment nextFragment,
                                                     FragmentTransaction fragmentTransaction) {
    }

    @Override
    public void applyCommands(Command[] commands) {
        fragmentManager.executePendingTransactions();

        //copy stack before apply commands
        copyStackToLocal();

        for (Command command : commands) {
            applyCommand(command);
        }
    }

    private void copyStackToLocal() {
        localStackCopy = new ArrayDeque<>();

        final int stackSize = fragmentManager.getBackStackEntryCount();
        for (int i = 0; i < stackSize; i++) {
            localStackCopy.push(fragmentManager.getBackStackEntryAt(i).getName());
        }
    }

    /**
     * Perform transition described by the navigation command
     *
     * @param command the navigation command to apply
     */
    protected void applyCommand(Command command) {
        if (command instanceof Forward) {
            forward((Forward) command);
        } else if (command instanceof Back) {
            back();
        } else if (command instanceof Replace) {
            replace((Replace) command);
        } else if (command instanceof BackTo) {
            backTo((BackTo) command);
        }
    }

    /**
     * Performs {@link Forward} command transition
     */
    protected void forward(Forward command) {
        Fragment fragment = createFragment(command.getScreenKey(), command.getTransitionData());

        if (fragment == null) {
            unknownScreen(command);
            return;
        }

        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        setupFragmentTransaction(
                command,
                fragmentManager.findFragmentById(containerId),
                fragment,
                fragmentTransaction
        );

        fragmentTransaction
                .replace(containerId, fragment)
                .addToBackStack(command.getScreenKey())
                .commit();
        localStackCopy.push(command.getScreenKey());
    }

    /**
     * Performs {@link Forward} command transition
     */
    protected void back() {
        if (localStackCopy.size() > 0) {
            fragmentManager.popBackStack();
            localStackCopy.pop();
        } else {
            exit();
        }
    }

    /**
     * Performs {@link Replace} command transition
     */
    protected void replace(Replace command) {
        Fragment fragment = createFragment(command.getScreenKey(), command.getTransitionData());

        if (fragment == null) {
            unknownScreen(command);
            return;
        }

        if (localStackCopy.size() > 0) {
            fragmentManager.popBackStack();
            localStackCopy.pop();

            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

            setupFragmentTransaction(
                    command,
                    fragmentManager.findFragmentById(containerId),
                    fragment,
                    fragmentTransaction
            );

            fragmentTransaction
                    .replace(containerId, fragment)
                    .addToBackStack(command.getScreenKey())
                    .commit();
            localStackCopy.push(command.getScreenKey());

        } else {
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

            setupFragmentTransaction(
                    command,
                    fragmentManager.findFragmentById(containerId),
                    fragment,
                    fragmentTransaction
            );

            fragmentTransaction
                    .replace(containerId, fragment)
                    .commit();
        }
    }

    /**
     * Performs {@link BackTo} command transition
     */
    protected void backTo(BackTo command) {
        String key = command.getScreenKey();

        if (key == null) {
            backToRoot();

        } else {
            if (localStackCopy.contains(key)) {
                Iterator<String> stackIterator = localStackCopy.iterator();
                while (!key.equals(stackIterator.next())) {
                    localStackCopy.pop();
                }
                fragmentManager.popBackStack(key, 0);
            } else {
                backToUnexisting(command.getScreenKey());
            }
        }
    }

    private void backToRoot() {
        fragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        localStackCopy.clear();
    }

    /**
     * Creates Fragment matching {@code screenKey}.
     *
     * @param screenKey screen key
     * @param data      initialization data
     * @return instantiated fragment for the passed screen key
     */
    protected abstract Fragment createFragment(String screenKey, Object data);

    /**
     * Called when we try to back from the root.
     */
    protected abstract void exit();

    /**
     * Called when we tried to back to some specific screen (via {@link BackTo} command),
     * but didn't found it.
     * @param screenKey screen key
     */
    protected void backToUnexisting(String screenKey) {
        backToRoot();
    }


    /**
     * Called if we can't create a screen.
     */
    protected void unknownScreen(Command command) {
        throw new RuntimeException("Can't create a screen for passed screenKey.");
    }
}

File: library/src/main/java/ru/terrakok/cicerone/android/FragmentNavigator.kt
package ru.terrakok.cicerone.android

import android.app.Fragment
import android.app.FragmentManager
import android.app.FragmentTransaction

import ru.terrakok.cicerone.Navigator
import ru.terrakok.cicerone.commands.Back
import ru.terrakok.cicerone.commands.BackTo
import ru.terrakok.cicerone.commands.Command
import ru.terrakok.cicerone.commands.CreationalCommand
import ru.terrakok.cicerone.commands.Forward
import ru.terrakok.cicerone.commands.Replace
import ru.terrakok.cicerone.commands.SystemMessage

/**
 * Implementation of [Navigator] based on fragments.
 *
 * [BackTo] navigation command will return to root if needed screen isn't found in the screens
 * chain. To change this behavior override [backToUnexisting] method.
 *
 * [Back] command will call [exit] method if current screen is root.
 *
 * @param fragmentManager fragment manager
 * @param containerId     id of the fragments container layout
 * @author Konstantin Tskhovrebov (aka terrakok) on 11.10.16.
 * @see SupportFragmentNavigator
 */
abstract class FragmentNavigator(
        private val fragmentManager: FragmentManager,
        private val containerId: Int
) : Navigator {

    override fun applyCommand(command: Command) {
        when (command) {
            is Forward -> applyForward(command)
            is Back -> applyBack()
            is Replace -> applyReplace(command)
            is BackTo -> applyBackTo(command)
            is SystemMessage -> showSystemMessage(command.message)
        }
    }

    private fun applyForward(forward: Forward) {
        val fragment = createFragment(forward.screenKey, forward.transitionData)
        if (fragment != null) {
            openFragment(forward, fragment)
        } else {
            unknownScreen(forward)
        }
    }

    private fun applyBack() {
        if (fragmentManager.backStackEntryCount > 0) {
            fragmentManager.popBackStackImmediate()
        } else {
            exit()
        }
    }

    private fun applyReplace(replace: Replace) {
        val fragment = createFragment(replace.screenKey, replace.transitionData) ?: run {
            unknownScreen(replace)
            return
        }

        if (fragmentManager.backStackEntryCount > 0) {
            fragmentManager.popBackStackImmediate()
            openFragment(replace, fragment)
        } else {
            openFragment(replace, fragment, addToBackStack = true)
        }
    }

    /**
     * Creates Fragment matching [screenKey].
     *
     * If it returns null, will be called [unknownScreen].
     *
     * @param screenKey screen key
     * @param data      initialization data, can be null
     * @return instantiated fragment for the passed screen key, or null if there no fragment that
     * accords to passed screen key
     */
    protected abstract fun createFragment(screenKey: String, data: Any?): Fragment?

    private fun openFragment(
            command: CreationalCommand,
            fragment: Fragment,
            addToBackStack: Boolean = true
    ) {
        fragmentManager.makeTransaction {
            setupAnimation(
                    command = command,
                    currentFragment = fragmentManager.findFragmentById(containerId),
                    nextFragment = fragment
            )

            replace(containerId, fragment)
            if (addToBackStack) addToBackStack(command.screenKey)
        }
    }

    /**
     * Called if we can't create a screen.
     */
    protected open fun unknownScreen(command: Command) {
        throw RuntimeException("Can't create a screen for passed screenKey.")
    }

    /**
     * Variant of [setupFragmentTransactionAnimation] where [FragmentTransaction] used as
     * receiver.
     *
     * @receiver fragment transaction
     * @param command             current navigation command. Will be only [Forward] or [Replace]
     * @param currentFragment     current fragment in container (for [Replace] command it will be
     * screen previous in new chain, NOT replaced screen)
     * @param nextFragment        next screen fragment
     * @see setupFragmentTransactionAnimation
     */
    protected fun FragmentTransaction.setupAnimation(
            command: CreationalCommand,
            currentFragment: Fragment,
            nextFragment: Fragment
    ) {
        setupFragmentTransactionAnimation(command, currentFragment, nextFragment, this)
    }

    /**
     * Override this method to setup custom fragment transaction animation.
     *
     * @param command             current navigation command. Will be only [Forward] or [Replace]
     * @param currentFragment     current fragment in container (for [Replace] command it will be
     * screen previous in new chain, NOT replaced screen)
     * @param nextFragment        next screen fragment
     * @param fragmentTransaction fragment transaction
     */
    protected open fun setupFragmentTransactionAnimation(
            command: CreationalCommand,
            currentFragment: Fragment,
            nextFragment: Fragment,
            fragmentTransaction: FragmentTransaction
    ) {
        @Suppress("DEPRECATION")
        setupFragmentTransactionAnimation(
                command as Command,
                currentFragment,
                nextFragment,
                fragmentTransaction
        )
    }

    // For backward compatibility
    @Deprecated("use variant of this function with CreationalCommand instead")
    protected open fun setupFragmentTransactionAnimation(
            command: Command,
            currentFragment: Fragment,
            nextFragment: Fragment,
            fragmentTransaction: FragmentTransaction
    ) {
        // Do nothing by default
    }

    private fun applyBackTo(backTo: BackTo) {
        backTo.screenKey?.let { key ->
            var hasScreen = false
            for (i in 0 until fragmentManager.backStackEntryCount) {
                if (key == fragmentManager.getBackStackEntryAt(i).name) {
                    fragmentManager.popBackStackImmediate(key, 0)
                    hasScreen = true
                    break
                }
            }

            if (!hasScreen) {
                backToUnexisting()
            }
        } ?: backToRoot()
    }

    /**
     * Called when we tried to back to some specific screen, but didn't found it.
     */
    protected open fun backToUnexisting() {
        backToRoot()
    }

    private fun backToRoot() {
        fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
    }

    /**
     * Shows system message.
     *
     * @param message message to show
     */
    protected abstract fun showSystemMessage(message: String)

    /**
     * Called when we try to back from the root.
     */
    protected abstract fun exit()
}
